mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp

```diff
@@ -554,6 +555,95 @@ static IntegerAttr wrapNumericMemorySpace(MLIRContext *ctx, unsigned space) {
   return IntegerAttr::get(IntegerType::get(ctx, 64), space);
 }
 
+/// Generates a symbol with 0-sized array type for dynamic shared memory usage,
+/// or uses existing symbol.
+LLVM::GlobalOp
+getDynamicSharedMemorySymbol(ConversionPatternRewriter &rewriter,
+                             gpu::DynamicSharedMemoryOp op,
+                             const LLVMTypeConverter *typeConverter,
+                             MemRefType memrefType, unsigned alignmentBit) {
+  std::optional<LLVM::GlobalOp> existingGlobalOp;
+
+  LLVM::LLVMFuncOp funcOp = op->getParentOfType<LLVM::LLVMFuncOp>();
+  assert(funcOp && "cannot find llvm.func op");
+
+  gpu::GPUModuleOp moduleOp = funcOp->getParentOfType<gpu::GPUModuleOp>();
+  assert(moduleOp && "cannot find gpu.module op");
+
+  // Use already generated global op if it exists
+  int index = 0;
+  std::string prefix = llvm::formatv("__shmem_{0}", funcOp.getSymName());
+  moduleOp->walk([&](LLVM::GlobalOp globalOp) {
+    if (auto arrayType = dyn_cast<LLVM::LLVMArrayType>(globalOp.getType())) {
+      if (arrayType.getNumElements() == 0) {
+        existingGlobalOp = globalOp;
+        return WalkResult::interrupt();
+      }
+    }
+    if (globalOp.getSymName().startswith(prefix))
+      index++;
```

Comment:
Why is this necessary? If this wants to avoid duplicate names, the code should just use `SymbolTable::insert` API instead that handles renaming correctly. This won't be correct for function names that share a prefix, e.g., `__shmem_foo` will be a prefix for `__shmem_foobar` when the latter clearly won't lead to a naming conflict.

Reply:
Agreed using SymbolTable! This was a workaround for name clashing. While building the symbol table with the GPUModule, I am getting the following errors. I will look deeper into this tomorrow, but if you can spot any obvious issues, please let me know 


```
gpu::GPUModuleOp moduleOp = funcOp->getParentOfType<gpu::GPUModuleOp>();
SymbolTable table(moduleOp);
```

```
mlir-opt: /usr/local/google/home/gurayozen/work/llvm-project/mlir/lib/IR/SymbolTable.cpp:137: mlir::SymbolTable::SymbolTable(mlir::Operation *): Assertion `inserted.second && "expected region to contain uniquely named symbol operations"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: mlir-opt dynamic-shared-memory.mlir -convert-gpu-to-nvvm -cse -canonicalize -o asd
----
#10 0x0000556d3e0c3e43 mlir::SymbolTable::SymbolTable(mlir::Operation*) /usr/local/google/home/gurayozen/work/llvm-project/mlir/lib/IR/SymbolTable.cpp:129:3
#11 0x0000556d3cbedf15 getDynamicSharedMemorySymbol(mlir::ConversionPatternRewriter&, mlir::gpu::DynamicSharedMemoryOp, mlir::LLVMTypeConverter const*, mlir::MemRefType, unsigned int) /usr/local/google/home/gurayozen/work/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:577:3
#12 0x0000556d3cbee362 mlir::GPUDynamicSharedMemoryOpLowering::matchAndRewrite(mlir::gpu::DynamicSharedMemoryOp, mlir::gpu::DynamicSharedMemoryOpAdaptor, mlir::ConversionPatternRewriter&) const /usr/local/google/home/gurayozen/work/llvm-project/mlir/lib/Conversion/GPUCommon/GPUOpsLowering.cpp:619:28
---
```

Reply:
I can't create a symbol table because the IRs in `gpu::GPUModuleOp` are messed up by the time when my patterns run. 

I search for unique names manually. What do you think?

Reply:
Messed up how? If they have duplicate names, they are not valid IR. We generally expect that IR is valid, at least at pass boundaries. I understand we may be inside a pass, but there may be a way to stage things even within one pass. Otherwise, I'd suggest to refactor the implementation of SymbolTable-based renaming into a utility function and use that.

Reply:
Here what I mean by that. The IR isn't fully lowered yet by the time this PR's patterns run, so I guess that's why I cannot build a symbol table (I get errors above) 
 
```
"gpu.module"() ({
  "llvm.func"() <{CConv = #llvm.cconv<ccc>, function_type = !llvm.func<void (i64)>, linkage = #llvm.linkage<external>, sym_name = "dynamic_shared_memory_kernel", visibility_ = 0 : i64}> ({
  ^bb0(%arg0: i64):
    %0 = "llvm.mlir.constant"() <{value = 8192 : index}> : () -> i64
    %1 = "arith.constant"() <{value = 8192 : index}> : () -> index
    %2 = "llvm.mlir.constant"() <{value = 16384 : index}> : () -> i64
    %3 = "arith.constant"() <{value = 16384 : index}> : () -> index
    %4 = "gpu.dynamic_shared_memory"() : () -> memref<?xi8, #gpu.address_space<workgroup>>
    %5 = "gpu.dynamic_shared_memory"() : () -> memref<?xi8, #gpu.address_space<workgroup>>
    %6 = "memref.view"(%4, %1) : (memref<?xi8, #gpu.address_space<workgroup>>, index) -> memref<32x64xf32, #gpu.address_space<workgroup>>
    "test.use.shared.memory"(%6) : (memref<32x64xf32, #gpu.address_space<workgroup>>) -> ()
    %7 = "memref.view"(%4, %3) : (memref<?xi8, #gpu.address_space<workgroup>>, index) -> memref<32x64xf32, #gpu.address_space<workgroup>>
    "test.use.shared.memory"(%7) : (memref<32x64xf32, #gpu.address_space<workgroup>>) -> ()
    "gpu.return"() : () -> ()
  }) {gpu.kernel, gpu.known_block_size = array<i32: 1, 1, 1>, gpu.known_grid_size = array<i32: 1, 1, 1>, nvvm.kernel} : () -> ()
  "gpu.func"() <{function_type = (index) -> ()}> ({
  }) {gpu.kernel, gpu.known_block_size = array<i32: 1, 1, 1>, gpu.known_grid_size = array<i32: 1, 1, 1>, sym_name = "dynamic_shared_memory_kernel", workgroup_attributions = 0 : i64} : () -> ()
  "gpu.module_end"() : () -> ()
}) {sym_name = "modules"} : () -> ()
```

Reply:
Your backtrace seems to indicate that you're creating the `mlir::SymbolTable` during pattern application in `matchAndRewrite`. Since the IR is not guaranteed to be a in a valid shape anymore at that point in time you're seeing that error.

I believe if you create the `mlir::SymbolTable` before application, either in the constructor of the pattern or passing it into the pattern, that it should circumvent this issue (as its internal map is initialized with the IR prior to any rewriting). This theoretically has its own issuses that should not be a problem here but I don't think we currently have the infrastructure to share symbol tabless between patterns, do we?

Reply:
I did not know that IR can be invalid during `matchAndRewrite`. 
In this case, I can create the symbol table in the pass, and pass it to the pattern.

Reply:
> Since the IR is not guaranteed to be a in a valid shape anymore at that point in time you're seeing that error.

seems strange to me: what is leading to invalid IR?

> In this case, I can create the symbol table in the pass, and pass it to the pattern.

creating a symbol table during the pattern application seems costly anyway.
But passing it in require some guarantee about nothing invalidating it in between: not clear to me how it is enforced?

Reply:
> > Since the IR is not guaranteed to be a in a valid shape anymore at that point in time you're seeing that error.
> 
> seems strange to me: what is leading to invalid IR?

I believe this is due to the dialect conversion framework only erasing converted operations after all rewrites have been performed. During rewriting, the module will contain `Symbol`s with identical names (the symbol op being converted + the new operation resulting from the symbol op), making the IR invalid during rewriting. 
This can be seen above in the generic IR @grypp posted here https://github.com/llvm/llvm-project/pull/71546#discussion_r1388087774 where there is both a `llvm.func` and `gpu.func` with the same `sym_name` and the cause of the assertion when constructing the symbol table each application.

Reply:
Let me summarize what I understand:
- Constructing the Symbol table isn't possible since IR may be invalid due to dialect conversion framework
- Symbol table construction is costly in the pattern
- Pre-constructing the Symbol table in the pass doesn't ensure no invalidations in between

Given these, is manually finding a unique symbol name like in this PR an acceptable approach?

Reply:
As an alternative - I can generate a `LLVM::GlobalOp` using SymbolTable in the Pass. Then, use it in the pattern. This way guarantees 1) SymbolTable is valid and we can use its Insert API, 2) Build SymbolTable once per `gpu.module`, so not expensive 3) Another pattern cannot generates a symbol with the same name. 

How does that sound?

Reply:
> My suggestion is to (1) collect all names of globals upfront in the pattern before the and store them in StringSet<>

If going this route: then lazy initialization and `DenseSet<StringAttr>` :)

Reply:
> As an alternative - I can generate a `LLVM::GlobalOp` using SymbolTable in the Pass. Then, use it in the pattern. This way guarantees

We are already creating a symbol table in the [passs](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp#L590). So there's no new overhead, besides maybe inserting a global, this would also make the DynamicShared pattern more efficient as the patter knows what symbol to use from the start.

Reply:
@ftynse I have implemented the way you proposed. 

> > As an alternative - I can generate a LLVM::GlobalOp using SymbolTable in the Pass. Then, use it in the pattern. This way guarantees
>
> We are already creating a symbol table in the [passs](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp#L590). So there's no new overhead with this approach, besides maybe inserting a global, this would also make the DynamicShared pattern more efficient as the pattern knows what symbol to use from the start.

Initially, I implemented it this way I proposed, but later, I made a change. The pattern is no longer be a self-sufficient lowering pattern; it would require a Pass to generate a `GlobalOp`. It complicates integration with other compilers based on MLIR, such as IREE ([check how they use these lowerings](https://github.com/openxla/iree/blob/86336293a8066b396537fae117d8549460cd85fd/compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToNVVM.cpp#L164)). If another compiler wants to use this pattern, it needs to generate GlobalOp in its repository, leading to some code replication. 

> If going this route: then lazy initialization and DenseSet<StringAttr>?

This data structure won't be large considering the expected low number of `GlobalOps` in the IR. I chose `StringSet` for managing small sets but I'm not as familiar with `DenseSet`